### ISSUE TITLE
Not compatible with ZSH Vi-mode

### ISSUE THREAD

**Issue by RobertAudi** [hchbaw/auto-fu.zsh#29]
I just added the following snippet to my `.zshrc` (taken from the Internet, I know!):

``` sh
# Vi-mode
bindkey -v

bindkey '^P' up-history
bindkey '^N' down-history
bindkey '^?' backward-delete-char
bindkey '^h' backward-delete-char
bindkey '^w' backward-kill-word
bindkey '^r' history-incremental-search-backward

function zle-line-init zle-keymap-select {
    VIM_PROMPT="%{$fg_bold[yellow]%} [% NORMAL]%  %{$reset_color%}"
    RPS1="${${KEYMAP/vicmd/$VIM_PROMPT}/(main|viins)/} %{$fg[red]%}$(rvm_info)%{$reset_color%}"
    zle reset-prompt
}

zle -N zle-line-init
zle -N zle-keymap-select
export KEYTIMEOUT=1
```

If I put this before auto-fu then auto-fu works but not the Vi-mode, and if I put it after auto-fu then Vi-mode works but not auto-fu...

Weird.

**Comment by hchbaw** [comment]
Hi, @AzizLight.

Sadly, no. The plain zsh’s `bindkey -v` is not well supported yet.
**But… Yes**, I use some kind of vi-ish modal editings.

Here is quick setup guides which is close to mine.
For those who really want to jump in to the configuration hell, “Wellcome to the real world (tm)”.

``` .sh
# compinit registers some too-emacs-ish-to-me keybindings in ~/.zcompdump
bindkey -e
autoload -Uz compinit; compinit

# Normal mode keybindings, please put some more.
bindkey -v
bindkey '^P' up-history

# auto-fu.zsh expects that `bindkey -e`, so change it before sourcing. (ugh!)
bindkey -e
bindkey $'\e' vi-cmd-mode
. ./zsh-syntax-highlighting/zsh-syntax-highlighting.zsh
. ./auto-fu.zsh/auto-fu.zsh
zle-line-init () auto-fu-init; zle -N zle-line-init;

# Normal-mode-ish setup.
VIM_PROMPT="[% NORMAL]%"
RPS1="$VIM_PROMPT"

my-reset-prompt-maybe () {
  # XXX: While auto-stuff is in effect,
  # when hitting <Return>, $KEYMAP becomes to `main`:
  # <Return> → `reset-prompt`(*) → `accept-line` → `zle-line-init`
  # → `zle-keymap-select` → `reset-promt` (again!)
  # Skip unwanted `reset-prompt`(*).
  ((auto_fu_init_p==1)) && [[ ${KEYMAP-} == main ]] && return

  # XXX: Please notice that `afu` is treated as Insert-mode-ish.
  RPS1="${${KEYMAP/vicmd/$VIM_PROMPT}/(main|viins|afu)/}"
  zle reset-prompt
}

zle-keymap-select () {
  auto-fu-zle-keymap-select "$@"
  my-reset-prompt-maybe
}
zle -N zle-keymap-select
```

Hope this helps.

TL;DR: Regarding to the `bindkey -v`

As my own understanding, `bindkey -v` activates the `viins` keymap and uses it as the initial keymap for the command line editing. The `viins` keymap has a keybinding such that `"^["` ⇒ `vi-cmd-mode` and `vi-cmd-mode` toggles the `vicmd` keymap. Zsh can do its vi-ish modal editing behavior by toggling these 2 keymaps as my understanding.

On the other hand, auto-fu.zsh’s auto-stuffs are actually implemented on top of the keymap `afu`. So, I wanted to toggle `afu` and `vicmd` keymaps when doing the modal editings. This job are done in the function `auto-fu-zle-keymap-select`.

If you tweak `zle-keymap-select` and `zle-line-init`, please remember that auto-fu.zsh also uses it.

**Comment by RobertAudi** [comment]
Hello @hchbaw 
Thanks a lot for the long answer! :+1: 

My ZSH skills are really poor, so I did not understand everything you explained, so my next question could be very stupid:

Isn't it possible to disable auto-fu when the `Esc` key is pressed and then enable it again when I press any command that would bring me into insert mode?

**Comment by hchbaw** [comment]
Ah, sorry.

> Isn't it possible to disable auto-fu when the Esc key is pressed and then enable it again when I press any command that would bring me into insert mode?

The above example actually accomplishes here, it's quite complitated though.
Doesn't it work for you?

**Comment by RobertAudi** [comment]
I don't know yet, I did not have the time to test your code. Will do in a bit and I'll let you know what happens

**Comment by RobertAudi** [comment]
@hchbaw Works for me! Thanks a lot :+1:

**Comment by hchbaw** [comment]
@AzizLight Thank you very much for reporting back!
